language-plutus-core/plutus-ir/Language/PlutusIR/TypeCheck/Internal.hs

```diff
@@ -0,0 +1,345 @@
+            -- we must also check that its result type is EXACTLY `[[TypeCon tyarg1] ... tyargn]`
+            when (funResultType nTy /= expectedConRes) .
+                throwing _TypeErrorExt $ MalformedDataConstrResType ann expectedConRes
+        expectedConRes = mkAppliedTyCon dt -- [[TypeCon tyarg1] ... tyargn]
+
+-- | For a single binding, returns all newly-declared variables next to their normalized types
+-- There is one implicitly-declared variable: the destructor.
+-- NB: [Input] Assumes that the input is globally-unique and any input subtypes are prior normalized
+-- NB: [Output] Uses renaming to preserve global uniqueness at the output, so the outputs is ready to be put into an env
+bindingNewVars :: MonadQuote m
+               => Normalized (Binding TyName Name uni ann)
+               -> m [(Name, Normalized (Type TyName uni ()))]
+bindingNewVars = unNormalized >>> \case
+  -- @effectfully: is it needed to rename the termbind's normalized-type? I don't think so.
+  TermBind _ _ (VarDecl _ n ty) _rhs ->
+      pure [(n, Normalized $ void ty)]
+
+  DatatypeBind _ dt@(Datatype ann _ tyargs desName vdecls) -> do
+      -- build the patternfunctor's "normalized" type and from it the "normalized" destructor
+      pf <- mkDatatypePatternFunctor ann dt
+
+      let -- build the normalized constructors
+          constrs = fmap (\(VarDecl _ constrName constrTy) -> (constrName, PIR.mkIterTyForall tyargs constrTy))
+                         vdecls
+          des = (desName, mkDestructorTy ann pf dt)
+          newStructors = des:constrs
+
+      -- rename all *-structors before we put them into the env, to make sure we preserve global uniqueness
+      forMOf (traversed._2) newStructors $
+          Normalized . void <<$>> rename
+
+  TypeBind{} -> pure [] -- typebinds don't introduce vardecls
+
+-- Helpers
+----------
+
+-- | A wrapper for safety: check a type to be *-kinded, and only then normalize it
+-- NB: Output is normalized (but not wrapped in Normalized because it is used inside a Binding (and binding is not a functor))
+checkStarNorm :: AsTypeError e term uni ann
+              => Type TyName uni ann
+              -> TypeCheckM uni e (Type TyName uni ann)
+checkStarNorm ty = unNormalized <$>
+                   normalizeTypeM ty <*
+                   checkKindM (typeAnn ty) ty (Type ())
+
+-- | Scope a typechecking computation with the given binding's newly-introducing type (if there is one)
+withNewType :: Binding TyName name uni () -> TypeCheckM uni e a -> TypeCheckM uni e a
+withNewType = \case
+    TypeBind _ (TyVarDecl _ tn k) _ -> withTyVar tn k
+    DatatypeBind _ (Datatype _ (TyVarDecl _ tn k) _ _ _) -> withTyVar tn k
+    TermBind{} -> id -- no type to introduce
+
+-- | warning: Does not no do conflict check
+withNewTypes :: Foldable f => f (Binding TyName name uni ()) -> TypeCheckM uni e a -> TypeCheckM uni e a
+withNewTypes = flip $ foldr withNewType
+
+-- | warning: Does not no do conflict check
```

Comment:
It shouldn't, so there's no need to warn about it. There *are* no conflicts with uniques.